File: 1658_min_operations/min_operations.cpp
/*
 * @Date: 2023-01-07
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2023-01-07
 * @FilePath: /algorithm/1658_min_operations/min_operations.cpp
 */

#include <cassert>
#include <numeric>
#include <vector>

using namespace std;

class Solution {
 public:
  int minOperations(vector<int>& nums, int x) {
    int n = nums.size();
    int sum = accumulate(nums.begin(), nums.end(), 0);

    if (sum < x) {
      return -1;
    }

    int right = 0;
    int lsum = 0, rsum = sum;
    int ans = n + 1;

    for (int left = -1; left < n; ++left) {
      if (left != -1) {
        lsum += nums[left];
      }
      while (right < n && lsum + rsum > x) {
        rsum -= nums[right];
        ++right;
      }
      if (lsum + rsum == x) {
        ans = min(ans, (left + 1) + (n - right));
      }
    }

    return ans > n ? -1 : ans;
  }
};

int main() {
  {
    vector<int> nums{1, 1, 4, 2, 3};
    int x = 5;
    int ans = 2;
    assert(Solution().minOperations(nums, x) == ans);
  }

  {
    vector<int> nums{5, 6, 7, 8, 9};
    int x = 4;
    int ans = -1;
    assert(Solution().minOperations(nums, x) == ans);
  }

  {
    vector<int> nums{3, 2, 20, 1, 1, 3};
    int x = 10;
    int ans = 5;
    assert(Solution().minOperations(nums, x) == ans);
  }
}

File: rust/2009_min_operations/min_operations.rs
/*
 * @Date: 2024-04-08
 * @LastEditors: 854284842@qq.com
 * @LastEditTime: 2024-04-08
 * @FilePath: /algorithm/rust/2009_min_operations/min_operations.rs
 */

struct Solution;

use std::collections::HashSet;

impl Solution {
    pub fn min_operations(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        let sorted_unique_nums: Vec<i32> = nums
            .iter()
            .cloned()
            .collect::<HashSet<_>>()
            .into_iter()
            .collect();
        let mut sorted_unique_nums = sorted_unique_nums;
        sorted_unique_nums.sort_unstable();
        let mut res = n as i32;
        let mut j = 0;
        for (i, &left) in sorted_unique_nums.iter().enumerate() {
            let right = left + n as i32 - 1;
            while j < sorted_unique_nums.len() && sorted_unique_nums[j] <= right {
                res = res.min(n as i32 - (j - i + 1) as i32);
                j += 1;
            }
        }
        res
    }
}

fn main() {
    let tests = vec![
        (vec![4, 2, 5, 3], 0),
        (vec![1, 2, 3, 5, 6], 1),
        (vec![1, 10, 100, 1000], 3),
    ];

    for (input, ans) in tests {
        assert_eq!(Solution::min_operations(input), ans);
    }
}
